### PR TITLE
Dima

### PR DESCRIPTION
**Pull request by DzmitryI**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,5 @@
 npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
+
+dataBases
```
Comment (maximsan) [inline .gitignore]: It's better to name you folder like 'data-bases' if you want to have compound name. It's the name convention.

File: src/components/memberPage/memberPage.js [hunk 2]
```diff
@@ -0,0 +1,378 @@
+import React, { Component } from 'react';
+import SwapiService from '../../services/swapi-service';
+import Backdrop from '../UI/backdrop';
+
+export default class MemberPage extends Component {
+  // constructor(props) {
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: remove

File: src/components/memberPage/memberPage.css [hunk 3]
```diff
@@ -0,0 +1,102 @@
+.memberWrap {
```
Comment (maximsan) [inline src/components/memberPage/memberPage.css]: It's better to give clear name like 'member-wrapper'

File: src/components/memberPage/memberPage.css [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  top: 0;
+  left: 0;
+  right: 0;
+}
+
+.close {
+  display: none;
+}
+
+.memberForm {
```
Comment (maximsan) [inline src/components/memberPage/memberPage.css]: It's better to name like 'member-form'

File: src/components/memberPage/memberPage.css [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+select {
+  width: 100%;
+  margin-right: 15px;
+  margin-left: 15px;
+  font-size: inherit;
+  outline: none;
+  box-sizing: border-box;
+}
+
+.btnAbb,
```
Comment (maximsan) [inline src/components/memberPage/memberPage.css]: You need to name it better. What is the idea ?

File: src/components/memberPage/memberPage.js [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+    isFormValid: false,
+    member: {
+      UserId: {
+        value: '',
+        // valid: false,
+        // touched: false,
+        // validation: {
+        //   required: true,
+        //   userId: true
+        // }
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: remove

File: src/components/memberPage/memberPage.js [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    if (!validation) {
+      return true;
+    }
+    let isValid = true;
+
+    if (validation.required) {
+      isValid = value.trim() !== '' && isValid;
+    }
+
+    return isValid;
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Make this method simpler.

File: src/components/memberPage/memberPage.js [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+  }
+
+  isInValid(valid, touched, shouldValidation) {
+    return !valid && touched && shouldValidation;
+  }
+
+  handleImput = ({ target }) => {
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Fix typo. Do target & state destruction at first line.

File: src/components/memberPage/memberPage.js [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+      UserId: {
+        value: '',
+        // valid: false,
+        // touched: false,
+        // validation: {
+        //   required: true,
+        //   userId: true
+        // }
+      },
+      Name: {
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: We do not use Uppercase name in javascript.

File: src/components/memberPage/memberPage.js [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+      member.MobilePhone.touched = true;
+      member.MobilePhone.valid = this.validateControl(target.value, member.MobilePhone.validation);
+    } else if (target.id === 'skype') {
+      member.Skype.value = target.value;
+      member.Skype.touched = true;
+      member.Skype.valid = this.validateControl(target.value, member.Skype.validation);
+    } else if (target.id === 'startDate') {
+      member.StartDate.value = target.value;
+      member.StartDate.touched = true;
+      member.StartDate.valid = this.validateControl(target.value, member.StartDate.validation);
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Please, make one common method. Follow DRY principle.

File: src/components/memberPage/memberPage.js [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+    return isValid;
+  }
+
+  isInValid(valid, touched, shouldValidation) {
+    return !valid && touched && shouldValidation;
+  }
+
+  handleImput = ({ target }) => {
+    const member = { ...this.state.member };
+    if (target.id === 'name') {
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: If you need to use string name, it's better to create a map object for it. By this you avoid possibility to typo.

File: src/components/memberPage/memberPage.js [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+        name !== 'DirectionId' &&
+        name !== 'UniversityAverageScore' &&
+        name !== 'MathScore' &&
+        name !== 'BirthDate' &&
+        name !== 'Sex'
+      ) {
+        isFormValid = member[name].valid && isFormValid;
+      }
+    });
+    console.log(member);
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: remove

File: src/components/memberPage/memberPage.js [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+    this.setState({ member: member, isFormValid: isFormValid });
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Use short object props. Reduce you method. It seems extremely big & ambiguous.

File: src/components/memberPage/memberPage.js [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+    } else if (target.id === 'startDate') {
+      member.StartDate.value = target.value;
+      member.StartDate.touched = true;
+      member.StartDate.valid = this.validateControl(target.value, member.StartDate.validation);
+    }
+
+    let isFormValid = true;
+    Object.keys(member).forEach((name) => {
+      if (
+        name !== 'UserId' &&
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: If you need to use string name, it's better to create a map object for it. By this you avoid possibility to typo.

File: src/components/memberPage/memberPage.js [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+      ) {
+        isFormValid = member[name].valid && isFormValid;
+      }
+    });
+    console.log(member);
+    this.setState({ member: member, isFormValid: isFormValid });
+  };
+
+  createMemberHandler = (event) => {
+    event.preventDefault();
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: it seems like extra action here

File: src/components/memberPage/memberPage.js [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+    this.setState({ member: {} });
+    this.props.onRegisterClick();
+  };
+
+  render() {
+    const {
+      Name: { value: Name },
+      LastName: { value: LastName },
+    } = this.state.member;
+    const cls = ['memberWrap'];
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: give it a clear name

File: src/components/memberPage/memberPage.js [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+    const clsInput = ['memberWrap'];
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: give it a clear name

File: src/components/memberPage/memberPage.js [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+    if (!this.props.isOpen) {
+      cls.push('close');
+    }
+    return (
+      <>
+        <div className={cls.join(' ')}>
+          <h1 className='title'>Create Member page</h1>
+          <form className='memberForm' method='post'>
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: actually, you don't need method here

File: src/components/memberPage/memberPage.js [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+            <h1 className='subtitle'>{`${Name} ${LastName}`}</h1>
+            <div className='formGroup'>
+              <label htmlFor='name'>Name</label>
+              <input type='text' id='name' onChange={this.handleImput} />
+              {this.isInValid(
+                this.state.member.Name.valid,
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Do destruction

File: src/components/memberPage/memberPage.js [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+            <div className='formGroup'>
+              <label htmlFor='lastName'>Last name</label>
+              <input type='text' id='lastName' onChange={this.handleImput} />
+              {this.isInValid(
+                this.state.member.LastName.valid,
+                this.state.member.LastName.touched,
+                this.state.member.LastName.validation.required,
+              ) ? (
+                <span>{this.state.member.LastName.errorMessage}</span>
+              ) : null}
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Please, extract this code to separate method. Follow DRY principle

File: src/components/memberPage/memberPage.js [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+            <div className='formGroup'>
+              <label htmlFor='email'>Email</label>
+              <input type='email' id='email' onChange={this.handleImput} />
+              {this.isInValid(
+                this.state.member.Email.valid,
+                this.state.member.Email.touched,
+                this.state.member.Email.validation.required,
+              ) ? (
+                <span>{this.state.member.Email.errorMessage}</span>
+              ) : null}
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Please, extract this code to separate method. Follow DRY principle

File: src/components/memberPage/memberPage.js [hunk 22]
```diff
@@ -0,0 +1,378 @@
...
+            <div className='formGroup'>
+              <label htmlFor='education'>Education</label>
+              <input type='text' id='education' onChange={this.handleImput} />
+              {this.isInValid(
+                this.state.member.Education.valid,
+                this.state.member.Education.touched,
+                this.state.member.Education.validation.required,
+              ) ? (
+                <span>{this.state.member.Education.errorMessage}</span>
+              ) : null}
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Please, extract this code to separate method. Follow DRY principle

File: src/components/memberPage/memberPage.js [hunk 23]
```diff
@@ -0,0 +1,378 @@
...
+                this.state.member.BirthDate.valid,
+                this.state.member.BirthDate.touched,
+                this.state.member.BirthDate.validation.required,
+              ) ? (
+                <span>{this.state.member.BirthDate.errorMessage}</span>
+              ) : null}
+            </div>
+            <div className='formGroup'>
+              <label htmlFor='universityAveraheScore'>University averahe score</label>
+              <input type='number' id='universityAveraheScore' onChange={this.handleImput} value='1' />
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: You should use defaultValue and uncontrollled component or value with setState. Don't write value directly.

File: src/components/memberPage/memberPage.js [hunk 24]
```diff
@@ -0,0 +1,378 @@
...
+              {this.isInValid(
+                this.state.member.BirthDate.valid,
+                this.state.member.BirthDate.touched,
+                this.state.member.BirthDate.validation.required,
+              ) ? (
+                <span>{this.state.member.BirthDate.errorMessage}</span>
+              ) : null}
+            </div>
+            <div className='formGroup'>
+              <label htmlFor='universityAveraheScore'>University averahe score</label>
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: You have typo in 'average'

File: src/components/memberPage/memberPage.js [hunk 25]
```diff
@@ -0,0 +1,378 @@
...
+              <input type='text' id='skype' onChange={this.handleImput} />
+              {this.isInValid(
+                this.state.member.Skype.valid,
+                this.state.member.Skype.touched,
+                this.state.member.Skype.validation.required,
+              ) ? (
+                <span>{this.state.member.Skype.errorMessage}</span>
+              ) : null}
+            </div>
+            <div className='formGroup'>
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: Create the FormGroup component and reuse it

File: src/components/memberPage/memberPage.js [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+            <div className='formGroup row'>
+              <button
+                className='btn btnAbb'
+                disabled={!this.state.isFormValid}
+                type='submin'
+                onClick={this.createMemberHandler}
+              >
+                Save
+              </button>
+              <button className='btn btnClose' onClick={this.props.onRegisterClick} type='button'>
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: rename class name to 'btn-close'

File: src/components/memberPage/memberPage.js [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+                Back to grid
+              </button>
+            </div>
+          </form>
+        </div>
+        {this.props.isOpen ? <Backdrop /> : null}
```
Comment (maximsan) [inline src/components/memberPage/memberPage.js]: do destruction

File: src/services/swapi-service.js [hunk 28]
```diff
@@ -0,0 +1,40 @@
+export default class SwapiService {
```
Comment (maximsan) [inline src/services/swapi-service.js]: rename it. this name is unclear isn't it ?

File: src/services/swapi-service.js [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+  _apiBase = `https://dims-f1a5f.firebaseio.com/`;
```
Comment (maximsan) [inline src/services/swapi-service.js]: you have to name constant value like API_BASE, it's javascript name convention

File: src/services/swapi-service.js [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+
+  getResource = async (url) => {
+    const res = await fetch(`${this._apiBase}${url}`);
```
Comment (maximsan) [inline src/services/swapi-service.js]: give it a full name

Comment (maximsan) [inline src/services/swapi-service.js]: if you use 'async await', you have to surround you code in 'try catch'

File: src/services/swapi-service.js [hunk 31]
```diff
@@ -0,0 +1,40 @@
...
+    if (!res.ok) {
+      throw new Error(`Could not fetch ${url}, received ${res.status}`);
+    }
+    return await res.json();
+  };
+
+  getAllPeople = async () => {
```
Comment (maximsan) [inline src/services/swapi-service.js]: why method names so different? Please, make them accurate

File: src/services/swapi-service.js [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+    const res = await this.getResource(`/UserProfile.json`);
+    const members = [];
+    Object.entries(res).forEach((key, value) => {
+      members.push({
+        value: key,
+      });
+    });
+    return members;
```
Comment (maximsan) [inline src/services/swapi-service.js]: try catch

File: src/services/swapi-service.js [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+  };
+
+  // _transformPerson = (person) => {
```
Comment (maximsan) [inline src/services/swapi-service.js]: remove

File: src/services/swapi-service.js [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+  setResource = async (url, body) => {
+    const res = await fetch(`${this._apiBase}${url}`, {
+      method: 'POST',
+      body: JSON.stringify(body),
+    });
+    return await res.json();
+  };
+
+  setPeople = async (body) => {
+    return await this.setResource(`/UserProfile.json`, body);
```
Comment (maximsan) [inline src/services/swapi-service.js]: try catch

File: src/services/swapi-service.js [hunk 35]
```diff
@@ -0,0 +1,40 @@
...
+  };
+
+  // _transformPerson = (person) => {
+  //   return {
+  //     userId: person.UserId,
+  //   };
+  // };
+
+  setResource = async (url, body) => {
+    const res = await fetch(`${this._apiBase}${url}`, {
```
Comment (maximsan) [inline src/services/swapi-service.js]: try catch

File: src/hoc/layout.jsx [hunk 36]
```diff
@@ -0,0 +1,11 @@
+import React, { Component } from 'react';
+
+export default class Layout extends Component {
```
Comment (maximsan) [inline src/hoc/layout.jsx]: make it stateless component

File: src/components/membersGrid/membersGrid.js [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+              <th>Full Name</th>
+              <th>Direction</th>
+              <th>Education</th>
+              <th>Start</th>
+              <th>Age</th>
+              <th></th>
+            </tr>
+          </thead>
+          <tbody>
+            {this.state.members.map((member, index) => {
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: Extract to renderMembers method

File: src/components/membersGrid/membersGrid.js [hunk 38]
```diff
@@ -0,0 +1,66 @@
...
+              return (
+                <tr key={member.value[0]}>
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: why we use here value[0] ? it's not a good idea

File: src/components/membersGrid/membersGrid.js [hunk 39]
```diff
@@ -0,0 +1,66 @@
...
+                  <td className='td'>{index + 1}</td>
+                  <td className='td'>{`${member.value[1].Name.value} ${member.value[1].LastName.value}`}</td>
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: why we use here value[1] ? it's not a good idea

File: src/components/membersGrid/membersGrid.js [hunk 40]
```diff
@@ -0,0 +1,66 @@
...
+export default class MembersGrid extends Component {
+  state = {
+    itemList: null,
+    members: [],
+  };
+
+  swapiService = new SwapiService();
+
+  componentDidMount() {
+    const getData = this.swapiService.getAllPeople;
```
Comment (maximsan) [inline src/components/membersGrid/membersGrid.js]: Why do we return Promise object here? You've created service for this.

**Comment by maximsan** [comment]
Please, look through notes and make sure it's clear for you. Enhance your code readability. Improve selectors, folders, files, methods naming. Re-do your pr. Thanks!
P.S. don't forget to follow DRY and KISS principles.

File: .gitignore (hunk 1 above)
Comment (DzmitryI) [reply .gitignore]: I deleted this file, because I don't need it

File: src/components/memberPage/memberPage.js (hunk 2 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.css (hunk 3 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.css]: ok

File: src/components/memberPage/memberPage.css (hunk 4 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.css]: ok

File: src/hoc/layout.jsx (hunk 36 above)
Comment (DzmitryI) [reply src/hoc/layout.jsx]: ok

File: src/components/membersGrid/membersGrid.js (hunk 37 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: ok

File: src/components/membersGrid/membersGrid.js (hunk 38 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: ok

File: src/components/membersGrid/membersGrid.js (hunk 39 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: ok

File: src/components/membersGrid/membersGrid.js (hunk 40 above)
Comment (DzmitryI) [reply src/components/membersGrid/membersGrid.js]: changed

File: src/components/memberPage/memberPage.js (hunk 6 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 9 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 8 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 10 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 12 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 13 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/services/swapi-service.js (hunk 33 above)
Comment (DzmitryI) [reply src/services/swapi-service.js]: ok

File: src/services/swapi-service.js (hunk 29 above)
Comment (DzmitryI) [reply src/services/swapi-service.js]: ok

File: src/components/memberPage/memberPage.js (hunk 27 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 26 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 25 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 15 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 16 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 17 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 18 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 19 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

File: src/components/memberPage/memberPage.js (hunk 20 above)
Comment (DzmitryI) [reply src/components/memberPage/memberPage.js]: ok

**Pull request closed without merging**